[Buffer Manager] Create volatile region for virtual memory management

nikriek opened the pull request:
This PR depends on https://github.com/hyrise/hyrise/pull/2618. VolatileRegion is used to create and manage the virtual memory region using various syscall e.g free(PageID) to call madvise(MADV_DONTNEED).

dey4ss committed:
Remove superfluous DISTINCT (#2561)

Implements a pattern that removes the DISTINCT clause if the result is unique by definition or previous GROUP BY clauses.

dey4ss committed:
Add Star Schema Benchmark (#2566)

Integrate the Star Schema Benchmark (SSB) as a Hyrise benchmark
executable

dey4ss committed:
Estimate cardinalities of predicates with uncorrelated subquery results (#2536)

Estimate query plans rewritten by the `JoinToPredicateRewriteRule` as if they were still (semi-) joins to place the rwritten predicates correctly in the query plans.

dey4ss committed:
Allow plan sharing for uncorrelated subqueries (#2555)

Treat uncorrelated subqueries just like normal inputs in Optimizer::validate_lqp(...)`.

dey4ss committed:
Limit scale factors for TPC-DS and SSB verification (#2573)

Ensure that the TPC-DS and SSB benchmark binaries crash when invoked with unfitting SFs and `--verify`.

dey4ss committed:
Add ProjectionNode before `SELECT DISTINCT` AggregateNode if necessary (#2571)

Handle `SELECT DISTINCT` statements correctly if there are further operations on the selected columns.

ClFeSc committed:
Use `std::stringstream` for default-type encoding message (#2581)

This PR uses a `std::stringstream` to encapsulate writing the
default-type encoding message to `std::cout` to prevent messages from
multiple threads being mangled.

dey4ss committed:
Support Clang-16 (#2583)

Small changes to support builds with clang 16.

niklasmohrin committed:
Make linker configurable and emit useful messages about linker choice (#2577)

Usage of linker LLD is not required anymore, but a warning is shown if the default linker is used on Linux. Added the option to use another linker by specifying `HYRISE_LINKER` (for example `mold`) - `CMAKE_LINKER` was actually already set to `/bin/ld`.

Bouncner committed:
Recursive Scheduler Test and Scheduler Shutdown Fix (#2580)

Adding a test that recursively adds tasks to the scheduler (a simple merge sort).

This test triggered a minor issue when shutting down the scheduler. This
issue should be fixed with the PR.

dey4ss committed:
Prepare subquery pruning (#2560)

This PR prepares the codebase for subquery pruning.

Bouncner committed:
No RowID temporaries for emplace_back() (#2585)

Simply replaces emplace_back(RowID{a,b}) with emplace_back(a,b)

Moreover, we increase the timeouts for the Python `pexpect` scripts as they often timed out when the load on the CI server was high.

Bouncner committed:
Update plot_performance_breakdown.py Script (#2587)

Updates the `plot_performance_breakdown.py` script to get rid of some warnings, generated PDF instead of PNG and improve legend placement.

SanJSp committed:
[DYOD] Team 3 - Sprint 4 (#2569)

Add tests for  `Difference` operator and the `<<` operator of the `OperatorScanPredicate`.

---------

Co-authored-by: Tarik Alnawa <43680469+T4rikA@users.noreply.github.com>
Co-authored-by: Paula <67064988+postmartem@users.noreply.github.com>
Co-authored-by: Tarik Alnawa <talnawa@outlook.de>
Co-authored-by: Bouncner <martin+github@boissier.de>
Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

niklasmohrin committed:
[DYOD] Team 1 - Sprint 4 (#2575)

Add tests for LQP nodes and operators.

Co-authored-by: Finn Schoellkopf <schoellkopf.finn@gmail.com>
Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

ClFeSc committed:
[DYOD] Team 2 - Sprint 4 (#2579)

Add tests for table and dummy table node.

judgeNotFound committed:
[DYOD] Team 4 - Sprint 4 (#2576)

Co-authored-by: Fabian Mahling <FabianMahling@web.de>
Co-authored-by: rothaarlappen <roessler.paul@web.de>
Co-authored-by: Fabian Mahling <45626086+Tratori@users.noreply.github.com>
Co-authored-by: Bouncner <martin+github@boissier.de>
Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

Bouncner committed:
Jenkins - Retry Mac x86 CI Runs (#2589)

This commit tells Jenkins to repeat the x86 macos step several times before failing.

j-hellenberg committed:
[DYOD] Team 5 - Sprint4 (#2578)

Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>
Co-authored-by: Bouncner <martin+github@boissier.de>

Bouncner committed:
Allow Apple Clang (#2584)

Enables Apple's clang version as it has sufficiently caught up with LLVM.

dey4ss committed:
Make CI runs more stable (#2594)

Fix CI issues with concurrent SSB data generation and `StoredtableNode`'s cached output expressions.

Co-authored-by: Bouncner <martin+github@boissier.de>

dey4ss committed:
Parse and represent window functions in the LQP (#2574)

Add the representation of SQL:2003 window functions in the LQP, including the translation in the SQLTranslator.

tjeyy committed:
Adapt IndexScan to use TableIndexes (#2553)

This PR adjusts the `IndexScan` to use the new `TableIndexes` introduced
in PR #2448 .

Co-authored-by: martin.boissier <martin.boissier@hpi.de>
Co-authored-by: Bouncner <martin+github@boissier.de>

Bouncner committed:
Scheduler: use different queue and semaphores for improved worker sleeping (#2548)

Using a new concurrent queue in the NodeQueueScheduler and semaphores instead of conditionals that workers wait on (less awaking, lower CPU usage when idling).

dey4ss committed:
Align TPC-C's Stock-Level with the standard's example procedure (#2604)

Replace the manually created `IN` list with a join, see #2602.

dey4ss committed:
Set max begin and end CIDs for each Insert/Delete (#2586)

Sets the maximal begin and end commit ID whenever an Insert or Delete operator commits.

dey4ss committed:
Switch off plan caching for benchmark runs with `--pipeline_metrics` (#2606)

Fixes performance and optimizer breakdown plots for cached queries.

dey4ss committed:
Use Predicates with Uncorrelated Subqueries for Dynamic Pruning (#2588)

Prunes tables using Predicates that contain uncorrelated subquery results. The information about such Predicates is added to StoredTableNodes by the ChunkPruningRule and later to GetTable operators by the LQPTranslator. Deep copies of LQPs and PQPs preserve the information.

dey4ss committed:
Add function prototypes in third-party DB generators (#2614)

Ensure C2x-compatibility, which clang-16+ warns about.

dey4ss committed:
Use TPC-C in Console test (#2615)

Replace TPC-DS with TPC-C, hoping for faster generation times without timeouts. Also, add pruning with between exclusive predicates to RangeFilter.

Bouncner committed:
Fix scheduler issues on NUMA servers (#2610)

No longer create TaskQueues for NUMA nodes without any workers and make scheduler shutdown mechanism more fail safe.

---------

Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

nikriek committed:
Add frame and page id

nikriek committed:
Add some more debug asserts

nikriek committed:
Add new line chars

nikriek committed:
Fix

nikriek committed:
Fix spaceship operator for now

nikriek committed:
Remove todo

nikriek committed:
Update order

nikriek committed:
Add basic implementation of volatile region

nikriek committed:
Fix order of imports again

nikriek committed:
Add additional tests

nikriek committed:
Empty-Commit

nikriek committed:
Fix tests

nikriek committed:
Add test for stream operators

nikriek committed:
Add missing part from codevcov

nikriek committed:
Apply some code comments

nikriek committed:
Update cpplint

nikriek committed:
Update many notes and apply comments

nikriek committed:
Update order

nikriek committed:
Add volatile region

nikriek committed:
Add more comments and reuse function on os x

Bouncner commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+
+Frame* VolatileRegion::get_frame(const PageID page_id) {
+  DebugAssert(page_id.size_type() == _size_type, "Page does not belong to this region.");
+  return &_frames[page_id.index()];
```
DebugAssert that we don't exceed the bounds?

Bouncner commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+  return sizeof(*this) + sizeof(decltype(_frames)::value_type) * _frames.capacity();
+}
+
+size_t VolatileRegion::size() const {
```
`frame_count`?

Bouncner replied:
page_count ... or something like that

Bouncner commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+
+void VolatileRegion::reuse(const PageID page_id) {
+  _unprotect_page(page_id);
+  // On OS X, we can use MADV_FREE_REUSE to update the memory accounting.
```
I don't remember where I noted that already, but there is no OS X anymore. It's called "macOS".

Bouncner commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+  }
+}
+
+inline std::size_t get_os_page_size() {
```
Reason so use `std::size_t` over `size_t` here?

Bouncner commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+    VolatileRegion::MappedRegion mapped_region) {
+  auto array = std::array<std::shared_ptr<VolatileRegion>, PAGE_SIZE_TYPES_COUNT>{};
+
+  // Ensure that every region has the same amount of virtual memory
```
Period.
Period.

Bouncner commented on src/lib/storage/buffer/volatile_region.hpp:
```diff
@@ -0,0 +1,105 @@
+class VolatileRegion final : public Noncopyable {
+ public:
+  // Default size of the virtual memory region for all page size types. Default is 256 GB.
+  constexpr static uint64_t DEFAULT_RESERVED_VIRTUAL_MEMORY = 1UL << 38;
```
`auto ... uint64_t{1} << 38` ?

I don't see any reason to use shifting here instead of writing the number here. But that might be my strange taste. :D

Bouncner replied:
What is the consequence of this size? What if we run SF 1000?

nikriek replied:
This would actually not work! We should think about this.

Bouncner commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+  for (auto i = size_t{0}; i < PAGE_SIZE_TYPES_COUNT; i++) {
+    auto region = MappedPageRegion{mapped_region.data() + DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION * i,
+                                   DEFAULT_RESERVED_VIRTUAL_MEMORY_PER_REGION};
+    array[i] = std::make_shared<VolatileRegion>(magic_enum::enum_value<PageSizeType>(i), region);
```
I haven't seen the usage yet, but I am surprised to see shared over unique pointers here.

Bouncner commented on src/lib/storage/buffer/volatile_region.hpp:
```diff
@@ -0,0 +1,105 @@
+  constexpr static bool ENABLE_MPROTECT = false;
+
+  // Create a VolatileRegion in a memory region for a given size_type.
+  VolatileRegion(const PageSizeType size_type, MappedPageRegion region);
```
Constructing the class and requiring to pass a data structure that is part of the class I am about to construct (MappedPageRegion) is strange.

Bouncner commented on src/lib/storage/buffer/volatile_region.hpp:
```diff
@@ -0,0 +1,105 @@
+  static void unmap_region(MappedRegion region);
+
+  // Create a VolatileRegion for each size type from the fixed-sized memory region.
+  static std::array<std::shared_ptr<VolatileRegion>, PAGE_SIZE_TYPES_COUNT> create_volatile_regions(
```
I think a little visualization or at least explanation in the header of what the connection between VolatileRegion, MappedRegion, and MappedPageRegion is, would be extremely helpful.

Bouncner reviewed (commented):
First round, haven't seen the tests yet.

Can you ping me when the diff is fixed?

dey4ss committed:
Parse and represent window functions in the LQP (#2574)

Add the representation of SQL:2003 window functions in the LQP, including the translation in the SQLTranslator.

dey4ss committed:
Make CI runs more stable (#2594)

Fix CI issues with concurrent SSB data generation and `StoredtableNode`'s cached output expressions.

Co-authored-by: Bouncner <martin+github@boissier.de>

nikriek committed:
Fix tests

dey4ss committed:
Prepare subquery pruning (#2560)

This PR prepares the codebase for subquery pruning.

dey4ss committed:
Remove superfluous DISTINCT (#2561)

Implements a pattern that removes the DISTINCT clause if the result is unique by definition or previous GROUP BY clauses.

nikriek committed:
Add test for stream operators

Bouncner committed:
Recursive Scheduler Test and Scheduler Shutdown Fix (#2580)

Adding a test that recursively adds tasks to the scheduler (a simple merge sort).

This test triggered a minor issue when shutting down the scheduler. This
issue should be fixed with the PR.

tjeyy committed:
Adapt IndexScan to use TableIndexes (#2553)

This PR adjusts the `IndexScan` to use the new `TableIndexes` introduced
in PR #2448 .

Co-authored-by: martin.boissier <martin.boissier@hpi.de>
Co-authored-by: Bouncner <martin+github@boissier.de>

judgeNotFound committed:
[DYOD] Team 4 - Sprint 4 (#2576)

Co-authored-by: Fabian Mahling <FabianMahling@web.de>
Co-authored-by: rothaarlappen <roessler.paul@web.de>
Co-authored-by: Fabian Mahling <45626086+Tratori@users.noreply.github.com>
Co-authored-by: Bouncner <martin+github@boissier.de>
Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

dey4ss committed:
Use Predicates with Uncorrelated Subqueries for Dynamic Pruning (#2588)

Prunes tables using Predicates that contain uncorrelated subquery results. The information about such Predicates is added to StoredTableNodes by the ChunkPruningRule and later to GetTable operators by the LQPTranslator. Deep copies of LQPs and PQPs preserve the information.

SanJSp committed:
[DYOD] Team 3 - Sprint 4 (#2569)

Add tests for  `Difference` operator and the `<<` operator of the `OperatorScanPredicate`.

---------

Co-authored-by: Tarik Alnawa <43680469+T4rikA@users.noreply.github.com>
Co-authored-by: Paula <67064988+postmartem@users.noreply.github.com>
Co-authored-by: Tarik Alnawa <talnawa@outlook.de>
Co-authored-by: Bouncner <martin+github@boissier.de>
Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

dey4ss committed:
Align TPC-C's Stock-Level with the standard's example procedure (#2604)

Replace the manually created `IN` list with a join, see #2602.

j-hellenberg committed:
[DYOD] Team 5 - Sprint4 (#2578)

Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>
Co-authored-by: Bouncner <martin+github@boissier.de>

dey4ss committed:
Allow plan sharing for uncorrelated subqueries (#2555)

Treat uncorrelated subqueries just like normal inputs in Optimizer::validate_lqp(...)`.

nikriek committed:
Add missing part from codevcov

dey4ss committed:
Estimate cardinalities of predicates with uncorrelated subquery results (#2536)

Estimate query plans rewritten by the `JoinToPredicateRewriteRule` as if they were still (semi-) joins to place the rwritten predicates correctly in the query plans.

nikriek committed:
Empty-Commit

dey4ss committed:
Add ProjectionNode before `SELECT DISTINCT` AggregateNode if necessary (#2571)

Handle `SELECT DISTINCT` statements correctly if there are further operations on the selected columns.

nikriek committed:
Fix order of imports again

niklasmohrin committed:
Make linker configurable and emit useful messages about linker choice (#2577)

Usage of linker LLD is not required anymore, but a warning is shown if the default linker is used on Linux. Added the option to use another linker by specifying `HYRISE_LINKER` (for example `mold`) - `CMAKE_LINKER` was actually already set to `/bin/ld`.

dey4ss committed:
Switch off plan caching for benchmark runs with `--pipeline_metrics` (#2606)

Fixes performance and optimizer breakdown plots for cached queries.

dey4ss committed:
Add Star Schema Benchmark (#2566)

Integrate the Star Schema Benchmark (SSB) as a Hyrise benchmark
executable

Bouncner committed:
No RowID temporaries for emplace_back() (#2585)

Simply replaces emplace_back(RowID{a,b}) with emplace_back(a,b)

Moreover, we increase the timeouts for the Python `pexpect` scripts as they often timed out when the load on the CI server was high.

nikriek committed:
Add additional tests

Bouncner committed:
Jenkins - Retry Mac x86 CI Runs (#2589)

This commit tells Jenkins to repeat the x86 macos step several times before failing.

dey4ss committed:
Support Clang-16 (#2583)

Small changes to support builds with clang 16.

Bouncner committed:
Update plot_performance_breakdown.py Script (#2587)

Updates the `plot_performance_breakdown.py` script to get rid of some warnings, generated PDF instead of PNG and improve legend placement.

nikriek committed:
Add more comments and reuse function on os x

Bouncner committed:
Scheduler: use different queue and semaphores for improved worker sleeping (#2548)

Using a new concurrent queue in the NodeQueueScheduler and semaphores instead of conditionals that workers wait on (less awaking, lower CPU usage when idling).

niklasmohrin committed:
[DYOD] Team 1 - Sprint 4 (#2575)

Add tests for LQP nodes and operators.

Co-authored-by: Finn Schoellkopf <schoellkopf.finn@gmail.com>
Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

dey4ss committed:
Set max begin and end CIDs for each Insert/Delete (#2586)

Sets the maximal begin and end commit ID whenever an Insert or Delete operator commits.

dey4ss committed:
Limit scale factors for TPC-DS and SSB verification (#2573)

Ensure that the TPC-DS and SSB benchmark binaries crash when invoked with unfitting SFs and `--verify`.

ClFeSc committed:
Use `std::stringstream` for default-type encoding message (#2581)

This PR uses a `std::stringstream` to encapsulate writing the
default-type encoding message to `std::cout` to prevent messages from
multiple threads being mangled.

ClFeSc committed:
[DYOD] Team 2 - Sprint 4 (#2579)

Add tests for table and dummy table node.

dey4ss committed:
Add function prototypes in third-party DB generators (#2614)

Ensure C2x-compatibility, which clang-16+ warns about.

Bouncner committed:
Allow Apple Clang (#2584)

Enables Apple's clang version as it has sufficiently caught up with LLVM.

nikriek committed:
Add volatile region

nikriek committed:
Update order

nikriek committed:
Update many notes and apply comments

nikriek committed:
Update cpplint

dey4ss committed:
Use TPC-C in Console test (#2615)

Replace TPC-DS with TPC-C, hoping for faster generation times without timeouts. Also, add pruning with between exclusive predicates to RangeFilter.

nikriek committed:
Apply some code comments

Bouncner committed:
Fix scheduler issues on NUMA servers (#2610)

No longer create TaskQueues for NUMA nodes without any workers and make scheduler shutdown mechanism more fail safe.

---------

Co-authored-by: Daniel Lindner <27929897+dey4ss@users.noreply.github.com>

nikriek committed:
Merge branch 'master' into nikriek/buffer-manager/volatile-region

nikriek committed:
Merge branch 'nikriek/buffer-manager/volatile-region' of https://github.com/hyrise/hyrise into nikriek/buffer-manager/volatile-region

nikriek committed:
Merge remote-tracking branch 'origin' into nikriek/buffer-manager/volatile-region

nikriek commented on src/lib/storage/buffer/volatile_region.hpp:
```diff
@@ -0,0 +1,96 @@
+ */
+class VolatileRegion final : public Noncopyable {
+ public:
+  // Enable mprotect class for debugging purposes
```
Makes sense. Debugging mode should be good.

nikriek commented on src/lib/storage/buffer/volatile_region.hpp:
```diff
@@ -0,0 +1,96 @@
+  // Returns the number of madvice(MADV_FREE_REUSABLE) calls on OS X or madvice(MADV_DONTNEED) calls on Linux
+  uint64_t madvice_free_call_count() const;
+
+  // Returns the number of page moevements using mbind or move_pages
```
LOL

nikriek commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,235 @@
+  }
+  _numa_page_movement_count.fetch_add(1, std::memory_order_relaxed);
+  _frames[page_id.index()].set_node_id(target_memory_node);
+#endif
```
Good point indeed.

nikriek commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,235 @@
+  return data;
+}
+
+Frame* VolatileRegion::get_frame(const PageID page_id) {
```
Good point

nikriek commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,235 @@
+  _madvice_free_call_count.fetch_add(1, std::memory_order_relaxed);
+}
+
+std::byte* VolatileRegion::get_page(const PageID page_id) {
```
I might want to use void*. We weed to work with raw memory

nikriek commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,238 @@
+  // The intermediate buffer only exists during the lifetime of the current thread.
+  // We create one buffer per thread to avoid contention. This buffer is divided into intermediate pages
+  // for each potential NUMA node. For simplicity, we use of the maximum page size to accommodate all page sizes.
+  static thread_local auto intermediate_buffer = []() {
```
Because you need to initialize a static thread-local value in several steps.

nikriek committed:
Apply some comments

nikriek commented on src/lib/storage/buffer/volatile_region.cpp:
```diff
@@ -0,0 +1,235 @@
+#if HYRISE_NUMA_SUPPORT
+  DebugAssert(target_memory_node != INVALID_NODE_ID, "Numa node has not been set.");
+  static thread_local std::vector<void*> pages_to_move{bytes_for_size_type(_size_type) / OS_PAGE_SIZE};
+  static thread_local std::vector<int> nodes{static_cast<int>(bytes_for_size_type(_size_type) / OS_PAGE_SIZE)};
```
Yes, but the move_pages syscall actually receives ints

tjeyy commented:
It looks like nobody is working on this PR right now.
